packages/core-modules/contracts/submodules/election/ElectionTally.sol

```diff
@@ -0,0 +1,109 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./ElectionBase.sol";
+
+contract ElectionTally is ElectionBase {
+    using SetUtil for SetUtil.AddressSet;
+
+    function _evaluateNextBallotBatch(uint numBallots) internal {
+        if (numBallots == 0) {
+            numBallots = _electionStore().settings.defaultBallotEvaluationBatchSize;
+        }
+
+        ElectionData storage election = _getCurrentElection();
+        uint totalBallots = election.ballotIds.length;
+
+        uint firstBallotIndex = election.numEvaluatedBallots;
+
+        uint lastBallotIndex = firstBallotIndex + numBallots;
+        if (lastBallotIndex >= totalBallots) {
+            lastBallotIndex = totalBallots - 1;
+
+            election.evaluated = true;
+        }
+
+        _evaluateBallotRange(election, firstBallotIndex, lastBallotIndex);
+    }
+
+    function _evaluateBallotRange(
+        ElectionData storage election,
+        uint fromIndex,
+        uint toIndex
+    ) internal {
+        ElectionSettings storage settings = _electionStore().settings;
+        uint numSeats = settings.nextEpochSeatCount;
+
+        for (uint ballotIndex = fromIndex; ballotIndex <= toIndex; ballotIndex++) {
+            bytes32 ballotId = election.ballotIds[ballotIndex];
+            BallotData storage ballot = election.ballotsById[ballotId];
+
+            _evaluateBallot(election, ballot, numSeats);
+        }
+    }
+
+    function _evaluateBallot(
+        ElectionData storage election,
+        BallotData storage ballot,
+        uint numSeats
+    ) internal {
+        uint ballotVotes = ballot.votes;
+        if (ballotVotes == 0) {
+            return;
+        }
+
+        address[] memory ballotCandidates = ballot.candidates;
+
+        uint numCandidates = ballotCandidates.length;
```

Comment:
Looks like is not necessary to assign the complete array from `storage` to `memory` on this use case:
```suggestion
        uint numCandidates = ballot.candidates.length;
```